src/SAWScript/Prover/MRSolver/Monad.hs

```diff
@@ -624,9 +695,49 @@ mrSubstEVarsStrict top_t =
 _mrSubstEVarsStrict :: Term -> MRM (Maybe Term)
 _mrSubstEVarsStrict = mrSubstEVarsStrict
 
+-- | Get the 'CoIndHyp' for a pair of 'FunName's, if there is one
+mrGetCoIndHyp :: FunName -> FunName -> MRM (Maybe CoIndHyp)
+mrGetCoIndHyp nm1 nm2 = Map.lookup (nm1, nm2) <$> mrCoIndHyps
+
+-- | Run a compuation under the additional co-inductive assumption that
+-- @forall x1, ..., xn. F y1 ... ym |= G z1 ... zl@, where @F@ and @G@ are
+-- the given 'FunName's, @y1, ..., ym@ and @z1, ..., zl@ are the given
+-- argument lists, and @x1, ..., xn@ is the current context of uvars. If
+-- while running the given computation a 'CoIndHypMismatchWidened' error is
+-- reached with the given names, the state is restored and the computation is
+-- re-run with the widened hypothesis. This is done recursively, meaning this
+-- function will only return once no 'CoIndHypMismatchWidened' errors are
+-- raised with the given names.
+withCoIndHyp :: FunName -> [Term] -> FunName -> [Term] -> MRM a -> MRM a
+withCoIndHyp nm1 args1 nm2 args2 m =
+  do ctx <- mrUVarCtx
+     withCoIndHyp' (nm1, nm2) (CoIndHyp ctx args1 args2) m
+
+-- | The main loop of 'withCoIndHyp'
+withCoIndHyp' :: (FunName, FunName) -> CoIndHyp -> MRM a -> MRM a
+withCoIndHyp' (nm1, nm2) hyp@(CoIndHyp _ args1 args2) m =
+  do mrDebugPPPrefixSep 1 "withCoIndHyp" (FunBind nm1 args1 CompFunReturn)
+                                    "|=" (FunBind nm2 args2 CompFunReturn)
+     st <- get
+     hyps' <- Map.insert (nm1, nm2) hyp <$> mrCoIndHyps
+     (local (\info -> info { mriCoIndHyps = hyps' }) m) `catchError` \case
+       CoIndHypMismatchWidened nm1' nm2' hyp' | nm1 == nm1' && nm2 == nm2'
+         -> -- FIXME: Could restoring the state here cause any problems?
+            put st >> withCoIndHyp' (nm1, nm2) hyp' m
```

Comment:
@eddywestbrook

Reply:
As long as the state only contains closed objects, it should be fine. That is, nothing in the state should contain any `Term`s that have any free deBruijn indices. The `MRVarMap` should always map top-level variables (i.e., `ExtCns`s) to info that contains only closed terms, so this should be the case.